### PR TITLE
Feat: Add Disco Programatic Issuance

### LINKED ISSUE #68: Bounty: Add Disco Programatic Issuance -  650 USDC + 50 OP Reward
# 👨‍💻 User Story
As a developer I want to easily issue verifiable credentials when users reach a specific goal or action.

More specifically I want to issue a "Proof of Hack" verifiable credential using the Disco credential issuance API.

# 📜 Disclaimer
To be eligible for the payout you must **first get approval** to work on the bounty.

- [Join the ⚡️TurboETH Discord channel](https://discord.gg/MKYh9pg8Yq)
- Introduce yourself in the `general` channel
- Announce your interest in the `bounty` channel

It's recommended to share a link to your personal Github account.

# 🧱 Project
Update the Disco integration to include support for verifiable credential programmatic issuance. 

You can reference the public [Disco schemas](https://github.com/discoxyz/disco-schemas) on Github for a full list of potential schemas 

https://github.com/discoxyz/disco-schemas

For the integration use the [ProofOfHackCredential](https://github.com/discoxyz/disco-schemas/tree/main/src/schemas/ProofOfHackCredential) verifiable credential.

https://github.com/discoxyz/disco-schemas/tree/main/src/schemas/ProofOfHackCredential

**Verifiable Credential Schema Snippet**
The core credential fields are "eventName", "eventDate" and "projectName", as denoted in the required field, but the form should include support for all of the properties.

```
credentialSubject: {
  type: "object",
  required: ["id", "eventName", "eventDate", "projectName"],
  properties: {
    id: { title: "Recipient DID", type: "string", format: "uri" },
    eventName: { title: "Event  Name", type: "string" },
    eventDate: { title: "Event Date", type: "string", format: "date" },
    projectName: { title: "Project Name", type: "string" },
    place: { title: "Place", type: "string" },
    teamName: { title: "Team Name", type: "string" },
    sourceCodeUrl: { title: "Source Code Url", type: "string" },
    usageLink: { title: "Usage Link", type: "string" },
  },
},
```

**Required Functionality:**
1. Disco verifiable credential issuance API integration
2. Form using [shacdn UI](https://ui.shadcn.com/docs/forms/react-hook-form) `<Form />` components.
3. New `app` page to issue verifiable credentials.

**Developer Tasks:**
- [ ] Create Hook - `disco/hooks/use-disco-issue-credential.tsx`
- [ ] Create API route - `disco/routes/route-credential-issue.tsx`
- [ ] Create Component - `disco/forms/form-credential-issuance-proof-of-hack.tsx`
- [ ] Create Page - `app/disco/proof-of-hack/page.tsx`

**Hook**
The `useDiscoIssueCredential` hook should use `useMutation` method exported from the [TanStack react-query](https://tanstack.com/query/v4/docs/react/reference/useMutation) package. 

**API**
The `routeCredentialIssuance.ts` endpoint should limit issuance to admins.

Admins are defined in `env` variables using the `APP_ADMINS` key.

The `isAdmin` key is added to authentication `session` during [account verification](https://github.com/turbo-eth/template-web3-app/blob/main/pages/api/siwe/verify.ts#L19).

https://github.com/turbo-eth/template-web3-app/blob/main/pages/api/siwe/verify.ts#L19

The `route` should be re-exported in a `pages/api/disco/credential-issue.ts` file.

https://github.com/turbo-eth/template-web3-app/tree/integrations/pages/api/disco

**Component**
The `<FormCredentialIssuanceProofOfHack />` component should use the `useDiscoIssueCredential` hook to interact with the API.

The form should be generated using the [shacdn UI component library](https://ui.shadcn.com/).

Follow the install instructions on https://ui.shadcn.com/docs/forms/react-hook-form

The form should look similar to the code below, but using the verifiable credential schema.

```
"use client"

import Link from "next/link"
import { zodResolver } from "@hookform/resolvers/zod"
import * as z from "zod"

import { Button } from "@/components/ui/button"
import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form"
import { Input } from "@/components/ui/input"

const formSchema = z.object({
  username: z.string().min(2, {
    message: "Username must be at least 2 characters.",
  }),
})

export function ProfileForm() {
. // 1. Define your form.
  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      username: "",
    },
  })
 
  // 2. Define a submit handler.
  function onSubmit(values: z.infer<typeof formSchema>) {
    // Do something with the form values.
    // ✅ This will be type-safe and validated.
    console.log(values)
  }
}

  return (
    <Form {...form}>
      <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8">
        <FormField
          control={form.control}
          name="username"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Username</FormLabel>
              <FormControl>
                <Input placeholder="shadcn" {...field} />
              </FormControl>
              <FormDescription>
                This is your public display name.
              </FormDescription>
              <FormMessage />
            </FormItem>
          )}
        />
        <Button type="submit">Submit</Button>
      </form>
    </Form>
  )
}
```

##### Integration
The [Disco](https://docs.disco.xyz/) integration is located in `integrations/disco` folder in the `integrations` branch on the `template-web3-app` repo.

```
integrations/disco
├─ routes/
│  ├─ route-credential-issue.tsx
├─ forms/
│  ├─ form-credential-issuance-proof-of-hack.tsx
├─ hooks/
│  ├─ **/*.ts
├─ utils/
│  ├─ **/*.ts
├─ client.ts
├─ README.md
```

##### Application
The application pages should be added to the `app/(general)/integration/disco` folder.
```
app/(general)/integration/disco
├─ proof-of-hack
│  ├─ page.tsx
```

# 💰 Bounty Reward
The bounty reward is 650 USDC and 50 OP tokens.

[TurboETH](https://www.turboeth.xyz/) is the recipient of [18,271.88 OP Tokens](https://mirror.xyz/kames.eth/pnWjQI3LmK5AThQTe9QiJyz-6b91lb61FMfk-dwrNyQ) from Optimism Retroactive Public Goods Funding. OP tokens earned from RPGF are helping fund TurboETH bounties.

### Notice
The final integration may not resemble the proposed integration - **that's O.K** - a natural part of software development.

During development you might discover an original hypothesis doesn't make sense. No problem. Make a comment and clearly explain why a new approach is better than old one. **Get rewarded for thinking out of the box.**

**The final bounty reward can be increased to match new bounty tasks.**

## Resources
- 🪩 Disco
	- Documentation - https://docs.disco.xyz
	- Application - https://app.disco.xyz
- ⚡️ TurboETH
	- Overview - http://docs.turboeth.xyz

### PR DESCRIPTION
**Pull request by 3scava1i3r**
Resolves #68 

Deliverables added  => 

- [x] Create Hook - disco/hooks/use-disco-issue-credential.tsx
- [x] Create API route - disco/routes/route-credential-issue.tsx
 - [x] Create Component - disco/forms/form-credential-issuance-proof-of-hack.tsx
 - [x] Create Page - app/disco/proof-of-hack/page.tsx
- [x] The <FormCredentialIssuanceProofOfHack /> component should use the useDiscoIssueCredential hook to interact with the API.
- [x] The form should be generated using the [shacdn UI component library](https://ui.shadcn.com/).
- [x] The useDiscoIssueCredential hook should use useMutation method exported from the [TanStack react-query](https://tanstack.com/query/v4/docs/react/reference/useMutation) package.
- [x] The routeCredentialIssuance.ts endpoint should limit issuance to admins.

Adding some screenshots for the same => 

image
image
image

### CODE REVIEW

File: integrations/disco/components/disco-issue-credentials-button.tsx [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import React from 'react'
```
Comment (marthendalnunes) [inline integrations/disco/components/disco-issue-credentials-button.tsx:1]:
```suggestion
```

File: integrations/disco/routes/get-profile-from-address/index.ts [hunk 2]
```diff
@@ -8,6 +8,7 @@ export async function discoGetProfileFromAddress(address?: string) {
       return null
     }
     const { data }: { data: Profile } = await discoClient.get(`/profile/address/${address}`)
+
```
Comment (marthendalnunes) [inline integrations/disco/routes/get-profile-from-address/index.ts:11]:
```suggestion
```

File: components/ui/input.tsx [hunk 3]
```diff
@@ -9,7 +9,7 @@ const Input = forwardRef<HTMLInputElement, InputProps>(({ className, ...props },
     <input
       ref={ref}
       className={cn(
-        'flex h-10 w-full rounded-md border border-slate-300 bg-transparent py-2 px-3 text-sm placeholder:text-slate-400 focus:outline-none focus:ring-2 focus:ring-slate-400 focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-700 dark:text-slate-50 dark:focus:ring-slate-400 dark:focus:ring-offset-slate-900',
+        'flex h-10 w-full rounded-md border border-slate-300 bg-transparent py-2 px-3 text-sm placeholder:text-slate-400 focus:outline-none focus:ring-2 focus:ring-slate-400 focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-400 dark:text-slate-50 dark:focus:ring-slate-400 dark:focus:ring-offset-slate-400',
```
Comment (marthendalnunes) [inline components/ui/input.tsx:12]: this change will affect all the input components over the entire repo, what's the reasoning behind it?

File: integrations/disco/forms/form-credential-issuance-proof-of-hack.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+const FormCredentialIssuanceProofOfHack = () => {
+  const { form, onSubmit, mutation } = useDiscoIssueCredential()
+
+  const { isLoading } = mutation
+
+  const { handleSubmit, register } = form
+
+  return (
+    <div className="card w-full">
+      <>
```
Comment (marthendalnunes) [inline integrations/disco/forms/form-credential-issuance-proof-of-hack.tsx:17]:
```suggestion
```

File: integrations/disco/forms/form-credential-issuance-proof-of-hack.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                />
+              )
+            })}
+
+            <Button className="btn btn-emerald w-full dark:bg-green-600 dark:text-white" disabled={isLoading}>
+              Issue
+            </Button>
+          </form>
+        </Form>
+      </>
```
Comment (marthendalnunes) [inline integrations/disco/forms/form-credential-issuance-proof-of-hack.tsx:50]:
```suggestion
```

File: integrations/disco/routes/post-credential-issue/client.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import { IssuedCredentials } from '../../utils/types'
+
+export async function appDiscoPostCredentialIssue(values: { recipientDid: any }) {
+  try {
+    const payload = {
+      schemaUrl: 'https://raw.githubusercontent.com/discoxyz/disco-schemas/main/json/ProofOfHackCredential/1-0-0.json',
+      subjectData: { ...values },
+      recipientDID: values?.recipientDid,
+    }
+    const { data }: { data: IssuedCredentials } = await axios.post(`/api/disco/credential-issue`, JSON.stringify(payload), {
```
Comment (marthendalnunes) [inline integrations/disco/routes/post-credential-issue/client.ts:12]: you can use `useQuery` from `tanstack-query` to handle the async state management

File: integrations/disco/routes/post-credential-issue/index.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import { IssuedCredentials } from '../../utils/types'
+
+export async function postCredentialIssue(val: { recipientDid: any; subjectData: any }) {
+  const payload = {
+    schemaUrl: 'https://raw.githubusercontent.com/discoxyz/disco-schemas/main/json/ProofOfHackCredential/1-0-0.json',
+    subjectData: { ...val?.subjectData },
+    recipientDID: val?.subjectData?.recipientDid,
+  }
+
+  const { data }: { data: IssuedCredentials } = await discoClient.post(`/credential`, JSON.stringify(payload), {
```
Comment (marthendalnunes) [inline integrations/disco/routes/post-credential-issue/index.ts:12]: you can use `useQuery` from `tanstack-query` to handle the async state management

**Comment by marthendalnunes** [comment]
@3scava1i3r the vercel deploy broke with the following message
image

File: components/ui/input.tsx (hunk 3 above)
Comment (3scava1i3r) [reply components/ui/input.tsx:12]: I was playing around with some additions there but it seems this was not reverted,I'll do the same asap

File: integrations/disco/routes/post-credential-issue/client.ts (hunk 6 above)
Comment (3scava1i3r) [reply integrations/disco/routes/post-credential-issue/client.ts:12]: for these i am currently using usemutate for statemanagement

File: integrations/disco/routes/post-credential-issue/index.ts (hunk 7 above)
Comment (3scava1i3r) [reply integrations/disco/routes/post-credential-issue/index.ts:12]: same as above

File: integrations/disco/hooks/use-disco-issue-credential.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
+import { setTimeout } from 'timers'
+
```
Comment (marthendalnunes) [inline integrations/disco/hooks/use-disco-issue-credential.ts:2]:
```suggestion
```

File: integrations/disco/hooks/use-disco-issue-credential.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    setTimeout(() => {
+      dismiss()
+    }, 4200)
+  }
+  const mutation = useMutation(async (vars) => {
+    const res = await appDiscoPostCredentialIssue(vars)
+
+    return res
+  })
```
Comment (marthendalnunes) [inline integrations/disco/hooks/use-disco-issue-credential.ts:29]: Always use the object input form for clarity, and the mutationFn expects a simple promise returning function, so there's no need to use `async await`. Also this usemutation should be abstracted to a hook like the other hooks in the integration
```suggestion
  const mutation = useMutation({
    mutationFn: (vars) => {
      return appDiscoPostCredentialIssue(vars)
    },
  })
```

File: integrations/disco/hooks/use-disco-issue-credential.ts
Comment (marthendalnunes) [inline integrations/disco/hooks/use-disco-issue-credential.ts:1]: I think this hook is misplaced here. It's a hook for the `form-credential-issuance-proof-of-hack.tsx` form. The other hooks are generic and meant tohave only disco api functionality. I recommend moving it elsewhere or adding a new hook to actually issue the credential wrapping appDiscoPostCredentialIssue into a usemutation like the rest of the hooks

File: integrations/disco/routes/post-credential-issue/client.ts (hunk 6 above)
Comment (marthendalnunes) [reply integrations/disco/routes/post-credential-issue/client.ts:12]: you're right

File: integrations/disco/routes/post-credential-issue/client.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+      headers: {
+        'Content-Type': 'application/json',
+      },
+    })
+    return data
+  } catch (error) {
+    throw error
+  }
```
Comment (marthendalnunes) [inline integrations/disco/routes/post-credential-issue/client.ts:20]: axios automatically handles errors, there's no need for a `try...catch` here

**Review by marthendalnunes — CHANGES_REQUESTED** [review]

File: integrations/disco/routes/post-credential-issue/client.ts (hunk 10 above)
Comment (3scava1i3r) [reply integrations/disco/routes/post-credential-issue/client.ts:20]: removing the same

File: integrations/disco/components/disco-issue-credentials-button.tsx
[the remark this answers is not in the document]
Comment (3scava1i3r) [reply integrations/disco/components/disco-issue-credentials-button.tsx:1]: deleted the same and added in pages section

File: integrations/disco/hooks/use-disco-issue-credential.ts (hunk 8 above)
Comment (3scava1i3r) [reply integrations/disco/hooks/use-disco-issue-credential.ts:2]: as handleToast is not being used,I'll remove the entire things

File: integrations/disco/hooks/use-disco-issue-credential.ts
Comment (3scava1i3r) [reply integrations/disco/hooks/use-disco-issue-credential.ts:1]: creating one for the form and other as usemutation return to be used later in the form hook as well

**Comment by marthendalnunes** [comment]
@3scava1i3r closing it since the #120 provides the same changes + a UI refactor on the integration

**Pull request closed without merging**